### ISSUE TITLE
Validator TypeError: Cannot read property 'allOf' of undefined

### ISSUE THREAD

**Issue by jimkoen** [cdimascio/express-openapi-validator#507]
I get this error when trying to validate my api spec with express-openapi-validator:

```
TypeError: Cannot read property 'allOf' of undefined
    at recurse (/root/web_push_api/node_modules/express-openapi-validator/dist/middlewares/parsers/schema.preprocessor.js:122:24)
    at /root/web_push_api/node_modules/express-openapi-validator/dist/middlewares/parsers/schema.preprocessor.js:144:21
    at Array.forEach (<anonymous>)
    at recurse (/root/web_push_api/node_modules/express-openapi-validator/dist/middlewares/parsers/schema.preprocessor.js:141:56)
    at SchemaPreprocessor.traverseSchemas (/root/web_push_api/node_modules/express-openapi-validator/dist/middlewares/parsers/schema.preprocessor.js:156:13)
    at SchemaPreprocessor.preProcess (/root/web_push_api/node_modules/express-openapi-validator/dist/middlewares/parsers/schema.preprocessor.js:65:14)
    at /root/web_push_api/node_modules/express-openapi-validator/dist/openapi.validator.js:81:95

```

To replicate this issue, clone [this repository](https://github.com/jimkoen/push_api_thing) and run `node main.js` and navigate to the /spec endpoint (literally using the code from the example).

The spec is validating without errors with `swagger-cli validate`

Notice that the definition works, when I define the schemas inline, eg. inside the yaml.

**Comment by jimkoen** [comment]
Edit: I needed to make the endpoints work, so I had to abandon trying to debug this. Initial debugging showed however, that the validator has issues dealing with a OpenAPI spec, which does not define a `components` section.

Inside [schema.preprocessor.ts](https://github.com/cdimascio/express-openapi-validator/blob/f466d9e971d2f64473fc3cf27aee924c0cbc016c/src/middlewares/parsers/schema.preprocessor.ts#L125) the function `gatherComponentSchemaNodes()` will fail, as it is expecting a schema with an existing components section.

I'm not familiar enough with the OpenAPI 3.0 Draft to make even an educated statement about this, but out of the three linters I used (spectral, ibm-lint-openapi, swagger-cli validation) all three report that my spec with the external jso-schema references is valid.

Notice however, that the bug seems to only occur sometimes, eg. the external json reference for the 'test' endpoint seems to be valid and accepted by the validator.

**Comment by cdimascio** [comment]
Thanks for the issue. A missing components section should be allowed. If it failed due to this, it’s likely a new bug. Which version of the validator are you running?

**Comment by cdimascio** [comment]
i added a [test](https://github.com/cdimascio/express-openapi-validator/blob/master/test/no.components.spec.ts#L37-L62) to check when components is not present. it passes

**Comment by cdimascio** [comment]
path resolution is the issue e.g. `#/paths/~1subscription/get/requestBody/...`
this is likely not an issue with `4.9.5`, you may wan to revert to that while i look into a fix

**Comment by cdimascio** [comment]
@jimkoen this should be fixed in `v4.10.2`. please reopen if you continue to have issues
